[flake] TestKubeadmControlPlaneReconciler_reconcileEtcdMembers fails at higher rate

### Which jobs are flaking?

periodic-cluster-api-test-mink8s-release-1-13	
periodic-cluster-api-test-main
periodic-cluster-api-test-mink8s-main	

### Which tests are flaking?

sigs.k8s.io/cluster-api/controlplane/kubeadm/internal/controllers TestKubeadmControlPlaneReconciler_reconcileEtcdMembers /Consider_alarms_from_external_members_when_computing_target_etcd_cluster_status


### Since when has it been flaking?

14th May 2026

### Testgrid link

https://storage.googleapis.com/k8s-triage/index.html?job=.*cluster-api&xjob=.*-provider-#1d29d51b4cf9bf479364

### Reason for failure (if possible)

Failed
=== RUN   TestKubeadmControlPlaneReconciler_reconcileEtcdMembers/Consider_alarms_from_external_members_when_computing_target_etcd_cluster_status
I0513 17:24:24.528470   29378 controller.go:1326] "Etcd members bar, foo without corresponding Machines, removing member bar"
I0513 17:24:24.528509   29378 remediation.go:697] "etcd cluster current state" currentTotalMembers=3 currentMembers=["m1-node","foo","bar"]
I0513 17:24:24.528602   29378 remediation.go:776] "An etcd member does not have a corresponding Machine, assuming member is healthy because it has no alarms" memberName="foo"
I0513 17:24:24.528659   29378 remediation.go:807] "etcd cluster considering removal of etcdMember bar" healthyMembers=["m1-node (m1)","foo (no machine)"] unhealthyMembers=[] totalMembers=2 totalLearnerMembers=0 totalVotingMembers=2 quorum=2 totalHealthyMembers=2 totalUnhealthyMembers=0 canSafelyTransitionToTargetState=true
I0513 17:24:24.528778   29378 controller.go:1342] "Etcd member without a corresponding Machine removed from the cluster" member="bar"
    controller_test.go:3501: 
        Expected an error to have occurred.  Got:
            <nil>: nil
--- FAIL: TestKubeadmControlPlaneReconciler_reconcileEtcdMembers/Consider_alarms_from_external_members_when_computing_target_etcd_cluster_status (0.00s)

### Anything else we need to know?

_No response_

### Label(s) to be applied

/kind flake
One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels.

--- a/controlplane/kubeadm/internal/controllers/controller.go
+++ b/controlplane/kubeadm/internal/controllers/controller.go
@@ -1318,7 +1318,9 @@ func (r *KubeadmControlPlaneReconciler) reconcileEtcdMembers(ctx context.Context
 		return ctrl.Result{}, errors.Wrap(err, "cannot get remote client to workload cluster")
 	}
 
-	etcdMemberToBeDeleted := unexpectedMembers.UnsortedList()[0]
+	sortedUnexpectedMembers := unexpectedMembers.UnsortedList()
+	sort.Strings(sortedUnexpectedMembers)
+	etcdMemberToBeDeleted := sortedUnexpectedMembers[0]
 	etcdMemberToBeDeletedMsg := etcdMemberToBeDeleted
 	if etcdMemberToBeDeletedMsg == "" {
 		etcdMemberToBeDeletedMsg = "(Name not yet assigned)"

--- a/controlplane/kubeadm/internal/controllers/controller_test.go
+++ b/controlplane/kubeadm/internal/controllers/controller_test.go
@@ -3458,7 +3458,33 @@ func TestKubeadmControlPlaneReconciler_reconcileEtcdMembers(t *testing.T) {
 			wantErr:                    true,
 		},
 		{
-			name: "Consider alarms from external members when computing target etcd cluster status",
+			name: "Consider alarms from members not being removed when computing target etcd cluster status",
+			controlPlane: &internal.ControlPlane{
+				KCP: &controlplanev1.KubeadmControlPlane{},
+				Machines: collections.Machines{
+					m1.Name: func() *clusterv1.Machine {
+						m := m1.DeepCopy()
+						conditions.Set(m, metav1.Condition{Type: controlplanev1.KubeadmControlPlaneMachineEtcdMemberHealthyCondition, Status: metav1.ConditionTrue, Reason: controlplanev1.KubeadmControlPlaneMachineEtcdMemberHealthyReason})
+						return m
+					}(),
+				},
+				EtcdMembersAlarms: []etcd.MemberAlarm{
+					{
+						MemberID: 1,
+						Type:     etcd.AlarmNoSpace,
+					},
+				},
+			},
+			additionalEtcdMembers: []*etcd.Member{
+				{Name: "foo", ID: 1},
+				{Name: "bar", ID: 2},
+			},
+			wantResult:                 ctrl.Result{},
+			wantRemoveEtcdMemberCalled: 0,
+			wantErr:                    true,
+		},
+		{
+			name: "Ignore alarms from external members being removed when computing target etcd cluster status",
 			controlPlane: &internal.ControlPlane{
 				KCP: &controlplanev1.KubeadmControlPlane{},
 				Machines: collections.Machines{
@@ -3471,7 +3497,7 @@ func TestKubeadmControlPlaneReconciler_reconcileEtcdMembers(t *testing.T) {
 				EtcdMembersAlarms: []etcd.MemberAlarm{
 					{
 						MemberID: 2,
-						Type:     etcd.AlarmNoSpace,
+						Type:     etcd.AlarmCorrupt,
 					},
 				},
 			},
@@ -3479,9 +3505,9 @@ func TestKubeadmControlPlaneReconciler_reconcileEtcdMembers(t *testing.T) {
 				{Name: "foo", ID: 1},
 				{Name: "bar", ID: 2},
 			},
-			wantResult:                 ctrl.Result{},
-			wantRemoveEtcdMemberCalled: 0,
-			wantErr:                    true,
+			wantResult:                 ctrl.Result{RequeueAfter: 1 * time.Second},
+			wantRemoveEtcdMemberCalled: 1,
+			wantErr:                    false,
 		},
 	}
 

--- a/controlplane/kubeadm/internal/controllers/remediation_test.go
+++ b/controlplane/kubeadm/internal/controllers/remediation_test.go
@@ -2380,29 +2380,64 @@ func TestTargetEtcdClusterHealthy(t *testing.T) {
 		g.Expect(canSafelyTransitionToTargetState).To(BeFalse())
 	})
 
-	t.Run("Can't safely remove a member from a CP with an etcd member without a machine but with alarms", func(t *testing.T) {
+	t.Run("Consider alarms from members not being removed when computing target etcd cluster status", func(t *testing.T) {
 		g := NewWithT(t)
 
 		m1 := getMachine(metav1.NamespaceDefault, "m1-mhc-unhealthy", withMachineHealthCheckFailed())
+		m2 := getMachine(metav1.NamespaceDefault, "m2-etcd-healthy", withHealthyEtcdMember())
 
 		controlPlane := &internal.ControlPlane{
-			Machines: collections.FromMachines(m1),
+			Machines: collections.FromMachines(m1, m2),
 		}
 		controlPlane.EtcdMembers = []*etcd.Member{
 			{
 				ID:   1,
-				Name: "member-without-machine",
+				Name: m1.Status.NodeRef.Name,
+			},
+			{
+				ID:   2,
+				Name: "foo",
 			},
 		}
 		controlPlane.EtcdMembersAlarms = []etcd.MemberAlarm{
 			{
-				MemberID: 1,
+				MemberID: 2,
 				Type:     etcd.AlarmNoSpace,
 			},
 		}
 
 		canSafelyTransitionToTargetState := r.targetEtcdClusterHealthy(ctx, controlPlane, false, m1.Status.NodeRef.Name)
 		g.Expect(canSafelyTransitionToTargetState).To(BeFalse())
+	})
+
+	t.Run("Ignore alarms from external members being removed when computing target etcd cluster status", func(t *testing.T) {
+		g := NewWithT(t)
+
+		m1 := getMachine(metav1.NamespaceDefault, "m1-mhc-unhealthy", withMachineHealthCheckFailed(), withHealthyEtcdMember())
+		m2 := getMachine(metav1.NamespaceDefault, "m2-etcd-healthy", withHealthyEtcdMember())
+
+		controlPlane := &internal.ControlPlane{
+			Machines: collections.FromMachines(m1, m2),
+		}
+		controlPlane.EtcdMembers = []*etcd.Member{
+			{
+				ID:   1,
+				Name: m1.Status.NodeRef.Name,
+			},
+			{
+				ID:   2,
+				Name: "bar",
+			},
+		}
+		controlPlane.EtcdMembersAlarms = []etcd.MemberAlarm{
+			{
+				MemberID: 2,
+				Type:     etcd.AlarmCorrupt,
+			},
+		}
+
+		canSafelyTransitionToTargetState := r.targetEtcdClusterHealthy(ctx, controlPlane, false, "bar")
+		g.Expect(canSafelyTransitionToTargetState).To(BeTrue())
 	})
 }
 

